src/services/history/HistoryService.ts

```diff
@@ -0,0 +1,126 @@
+import { useCanvasStore } from '@/store/CanvasStore'
+import {
+  useHistoryStore,
+  type ShapeDrawnEvent,
+  HistoryEventType,
+  type ShapeDeletedEvent,
+  type HistoryEvent,
+  type ShapesReplacedEvent,
+  type ShapeUpdatedEvent,
+  type AggragateEvent
+} from '@/store/HistoryStore'
+
+class HistoryService {
+  public undo() {
+    const historyStore = useHistoryStore()
+    const lastEvent = historyStore.popEvent()
+    if (!lastEvent) return
+    this.undoEvent(lastEvent)
+  }
+
+  public redo() {
+    const historyStore = useHistoryStore()
+    const revertedEvent = historyStore.unpopEvent()
+    if (!revertedEvent) return
+    this.redoEvent(revertedEvent)
+  }
+
+  private undoEvent(event: HistoryEvent) {
+    switch (event.type) {
+      case HistoryEventType.SHAPE_DRAWN:
+        this.undoShapeDrawnEvent(event)
+        break
+      case HistoryEventType.SHAPE_DELETED:
+        this.undoShapeDeletedEvent(event)
+        break
+      case HistoryEventType.SHAPES_REPLACED:
+        this.undoShapesReplacementEvent(event)
+        break
+      case HistoryEventType.SHAPE_UPDATED:
+        this.undoShapeUpdatedEvent(event)
+        break
+      case HistoryEventType.AGGREGATE:
+        this.undoAggregateEvent(event)
+        break
+    }
+  }
```

Comment:
This whole event bussiness thing feels like it deserves its own class hierachy. Not sure how that would work out in practice though...

Reply:
Not sure what you mean by that. Each event has more or less unique set of information that it needs to save to be reversable and I think it's rather convenient that TypeScript even allows to easily express such polymorphism in the first place

Reply:
Think on that!

Reply:
I thought about about that, and you are right! I will rewrite that so events will be objects that will know how to do and redo them